### ISSUE TITLE
Live config reload doesn't notice when symlink is replaced

### ISSUE THREAD

**Issue by Nemo157** [alacritty/alacritty#2237]
I use a config file manager that symlinks in my alacritty config file and replaces the symlink when updated. Because alacritty [canonicalizes the config filename](https://github.com/jwilm/alacritty/blob/a9c6e2807f3a205ee1adc8adb732aac9215ea6cb/src/config/mod.rs#L2298) it ends up watching the old config file and doesn't notice when the symlink is updated to point at the new config file (at least, I'm assuming this is the issue, I'll try to get around to adding some debugging and checking exactly what alacritty does soon).

It seems like accurately watching a symlinked file is going to be difficult, you need to notice when the symlink _or_ the targeted file changes, and there could be many layers of symlinks to watch (include having directory components symlinked and needing to notice when they update). If you feel this is too complex to be worth it I'm fine with marking this as won't fix.

(Linux 5.0.4-arch1-1-ARCH x86_64 GNU/Linux, X11)

**Comment by chrisduerr** [comment]
I think we could potentially make this a little better by checking if the real path of the file is different than the file itself and then watch both directories if that's the case. With nested symlink levels that would still only cover changing the file itself and the source file, but I think that's the only sane solution that doesn't involve a change on the user's system.

I'm not sure how watching symlinks works out with the library we're using though. This would need to be investigated first and we might need to consider that different kernels (especially linux/macos/bsd in this regard) might have different APIs to watch files used by our dependency.

Then the alternative is to change the user's system. Of course we can't really control this, but looking at your usecase, I feel like maybe hardlinks might help? The advantage of hardlinks is that they still get updated when either file is updated, however the real path does not resolve to the 'original file' (since there really is none). I think these should also work correctly with the library we're using and hopefully resolve all your problems?

**Comment by Nemo157** [comment]
I've actually investigated this slightly more and found the issue is slightly different to what I thought. The symlink is not being replaced, but a grandparent directory of the file it symlinks to is. The [tool I use](https://github.com/freshshell/fresh) creates a new directory, builds new versions of config files into it, then replaces a the grandparent with the new directory.

Adding some debugging it appears that alacritty only sees the removal of the config file and its direct parent once the new directory is in place, since it is not watching the grandparent directory (`/home/wim/.fresh/build`) that gets replaced:

```
[src/config/mod.rs:2310] rx.recv().expect("watcher event") = Remove(
    "/home/wim/.fresh/build/config-alacritty/alacritty.yml"
)
[src/config/mod.rs:2310] rx.recv().expect("watcher event") = Remove(
    "/home/wim/.fresh/build/config-alacritty"
)
```

---

A solution I see is to have alacritty watch each path component back to the root and filter for the events that could have caused the config to change, specifically in this case it also needs to handle

```
[src/config/mod.rs:2311] rx.recv().expect("watcher event") = Rename(
    "/home/wim/.fresh/build.new",
    "/home/wim/.fresh/build"
)
```

I think this should be relatively simple to add, if you're happy with that plan I can probably open a PR soon.

**Comment by chrisduerr** [comment]
Am I understanding correctly that you're proposing to watch every file in the symlink chain? That seems like a poor choice and excessive on resources.

Though since you said this should be 'relatively simple to add', I'd assume you're talking about something different?

**Comment by Nemo157** [comment]
@chrisduerr to watch every directory betweeen the file and the root. The only one in that path that would normally get a lot of activity is the home directory, but even that shouldn’t get enough to be a real issue.

**Comment by chrisduerr** [comment]
> @chrisduerr to watch every directory betweeen the file and the root. The only one in that path that would normally get a lot of activity is the home directory, but even that shouldn’t get enough to be a real issue.

That sounds like an extremely expensive hack just to compensate for something that can be fixed on the user's system by working around it.

That doesn't sound like a solution I'd be willing to accept into Alacritty.

**Comment by dnkl** [comment]
A possible solution would be to start polling for the configuration file *after you have seen it disappear*.

Under normal operation, be it *with* or *without* a configuration file, there would be no difference compared to now. But once you see the file disappear, start polling for it. Once it's back, start monitoring as usual (no polling). I think even something like 30s, or even 60s poll cycle would be acceptable.

**Comment by chrisduerr** [comment]
I'd be perfectly fine with that solution. As long as we don't have a super high polling rate there shouldn't be any problem and if the code is kept clean and straight-forward it should just be mostly positive.

However, assuming that the polling rate is relatively high (let's say 1s), it's possible that people would quit Alacritty before it has a chance to pick up the new file. So we need to strike a balance between useful reload time and saving resources here.

I'd also like to note that I definitely would want the behavior of keeping the current config loaded even when the file was deleted. So we shouldn't unload to the default config first and then reload whenever we find a config again.

**Comment by laurentS** [comment]
For what it's worth, I've been having a similar issue (use case: I want to color sets, one for day time, one for night time) where I tried to delete/recreate a symlink to the config file. The solution for me was to force/overwrite the symlink (using `ln -sf ...`). The same happened for hard links, the change was only picked up when overwriting the link. I don't know if this helps the situation described here, but just in case...

**Comment by lifranc** [comment]
Polling? No; Always push, never pull. So alacritty needs an IO socket. Check out [bincode](https://github.com/servo/bincode). Then you dont even need to include all this kludge in alacritty, like yaml parsing and the ohh-so-obsolete UnixFs watching.

**Comment by oknozor** [comment]
Hello, anyone is working on what  @dnkl proposed ? I would need this feature for the dotfile manager I am working on. If no one has started to work on it, I'll be glad to try, but I would need some guidance.

**Comment by maisiliym** [comment]
>  alacritty needs an IO socket.
Exactly; a modern program can communicate with the outside world.

**Comment by ncfavier** [comment]
This is partially fixed by #5399

**Comment by chrisduerr** [comment]
I think this issue is fixed about as much as it's ever gonna be.
